src/tracedAtomic.ml

```diff
@@ -259,6 +286,200 @@ let do_run init_func init_schedule =
+              in
+
+              (* List.iter
+                 (fun proc ->
+                   Printf.printf
+                     "[depth=%d]     ::occurs_after_e_but_indep:: (p=%d, %s)\n%!"
+                     depth proc.run_proc
+                     (atomic_op_str proc.run_op))
+                 v_E; *)
+              let v = v_E @ [ state_top ] in
+
+              (* List.iter
+                 (fun proc ->
+                   Printf.printf "[depth=%d]     ::v:: (p=%d, %s)\n%!" depth
+                     proc.run_proc
+                     (atomic_op_str proc.run_op))
+                 v; *)
+              let initials =
+                let initials_map, initials_spawns =
+                  List.fold_left
+                    (fun (first_accesses, spawns) (state_cell : state_cell) ->
+                      match state_cell.run_ptr with
+                      | None ->
+                          let new_spawns =
+                            IntSet.add state_cell.run_proc spawns
+                          in
+                          (first_accesses, new_spawns)
+                      | Some obj_ptr ->
+                          let new_first_accesses =
+                            IntMap.update obj_ptr
+                              (function
+                                | Some v -> Some v
+                                | None -> Some (`Run_proc state_cell.run_proc))
+                              first_accesses
+                          in
+                          (new_first_accesses, spawns))
+                    (IntMap.empty, IntSet.empty)
+                    v
+                in
+                IntMap.fold
+                  (fun _obj_ptr (`Run_proc run_proc) initials_set ->
+                    IntSet.add run_proc initials_set)
+                  initials_map initials_spawns
+              in
+              (*
+                           Printf.printf
+                             "[depth=%d]     initials=[%s], backtrack=[%s] at depth=%d, p=%d, \
+                              op=%s \n\
+                              %!"
+                             depth (_set_to_str initials)
+                             (_set_to_str prefix_top.backtrack)
+                             (List.length prefix - 1)
+                             prefix_top.run_proc
+                             (atomic_op_str prefix_top.run_op); *)
+              let prefix_top_sleep =
+                !(List.nth sleep_sets (List.length prefix - 1))
+              in
+              if
+                IntSet.(
+                  cardinal
+                    (inter prefix_top.backtrack
+                       (diff initials prefix_top_sleep)))
+                = 0
+              then
+                let initial =
+                  match
+                    IntSet.(min_elt_opt (diff initials prefix_top_sleep))
+                  with
+                  | Some initial -> initial
+                  | None -> IntSet.min_elt initials
```

Comment:
>Hm so I don't understand how there can be multiple initials? 

There's many ways to reverse a race and fewer alternatives where possible. Say, we are in a sequence E.t.w.t', where E, w are some sequences of transitions, and t, t' are single transitions, which constitute a reversible race (act on the same variable, belong to different proc). The strongest way to execute the reversal of t, t' is to add t' to E.backtrack. This guarantees that E.t' will be or has been explored. 

But that's not the only way. Suppose w is a single operation that is independent with all of our E, t, t' business. A spawn or just an access of a different var on a different thread. With that, we can explore the reversal as E.w.t' as well. Therefore, there are two initials = {t', w}.

Reply:
Ha yes, you are right, I forgot that I was thinking of the shortest (strongest) sequence which is unique. So in your example, I don't think we need to consider the initial `w`?

Reply:
We want to consider both initials to explore fewer interleavings. That's because the sequences E.w.t' and E.t'.w are equivalent and, generally, if we explore one initial, then we have explored both. I haven't penetrated the proof of it fully but it makes sense intuitively - throwing an indep operation in front of a racy program should not impact the result. 

We're not seeing this above, without the bias patch, because our computation of indep and initials is too simplistic (essentially more or less forgetting about transitivity of happens-before). We end up with too many initials, and some of them belong to traces, rather than the one we want to schedule.

A curious observation: with intransitive happens-before the without-bias version is fine on even very complicated two-thread tests, but it's really easy to break it with 3+. That's the smallest offending test case I found: 

```
let test () =
  let a = Atomic.make 0 in
  let b = Atomic.make 0 in

  Atomic.spawn (fun () ->
      Atomic.set b 1;
      Atomic.set a 2);

  Atomic.spawn (fun () -> Atomic.set a 4);

  Atomic.spawn (fun () -> Atomic.set b 8);
```

I have a patch that fixes the computation of indep and initials. I'll clean it up and push next week.

Reply:
Yeah so with two domains, one just needs to mark the backtrackable steps as there's only one alternative domain to choose from (so anything works ^^). With more domains, you can have a complex situation where backtracking one domain operation requires running (or not running) other domains first. The source set paper you linked has an example at the top of page 7, with 4 domains, to illustrate why wake-up trees are necessary for optimality.

In any case, sleeping sets are already pretty good, but:
- If the strongest initial is in the sleeping set, then one shouldn't attempt to backtrack there as it was already explored
- If the chosen initial is not in the enabled set at that time, then backtracking will silently ignore the branch

Pending your next fix, without the bias, I think you are in the second situation where the strongest initial is not correctly computed for some backtrackable steps (another domain should have been picked as initial, which would have enabled the branch). With the bias, I think you end up scheduling more branches since it doesn't actually check that the operation was backtrackable at that time (but as long as the enabled set is not considered, then you can still pick a random initial that won't be able to run).

Reply:
> Yeah so with two domains, one just needs to mark the backtrackable steps as there's only one alternative domain to choose from (so anything works ^^). 

Yep, basically there's no transitive case really. Something to be mindful of in the future.